Thio Leng Kiat iP

Thiolk opened the pull request.

Thiolk committed:
Level-0: Greet

Thiolk committed:
Add horizontal line method and implement in level 0

Thiolk committed:
Remove Line() (line 14) and add in printHorizontalLine method at the end

Thiolk committed:
Imported scanner class to receive user input

Thiolk committed:
Remove redundant return statement in printHorizontalLine method

Thiolk committed:
Implemented checkInput method so the bot knows how to respond
to the different inputs (echo it or respond with bye message)

Thiolk committed:
Added 2 new methods
* addToList
* listInputs

addToList allows the bot to store and track user's inputs
listInputs have yet to be implemented

checkInputs method currently only checked for 2 kinds of inputs

The current level requires to bot to be able to perform 3 actions

if else statements have been added to check for the 3 kinds of inputs

Thiolk committed:
Revert "Added 2 new methods"

This reverts commit 231df5770ef7d9de02cadadb7fa280b2ade00fa4.

Thiolk committed:
Implement addToList method and update checkInput method

The current bot can only echo user inputs

Current level requires bot to store user's inputs

addToList method allows the bot to store user inputs in an ArrayList

ArrayList was chosen as the list can be resized easily

checkInputs method currently only checked for 2 kinds of inputs

The current level requires to bot to be able to perform 3 actions

if else statements have been added to check for the 3 kinds of inputs

Thiolk committed:
Complete Level-2

Current bot can only store user's inputs.

To complete Level-2 requirements, bot needs to list all user inputs.

listInputs method implemented to list all user inputs using a for loop.

Thiolk committed:
Rename inputs to task

Bot acts as a to do list.

Renaming inputs to task where appropriate
increases code readability.

Thiolk committed:
Create Task Class

Tasks were managed by methods in Main.

A task class is created to manage all tasks as objects.

Creating a new class allows for more abstraction and follows the convention of OOP.

Thiolk committed:
Implement Task class in Level-1 code

Methods were used to handle the operations of tasks.

Task class is created and used to manage the operations of task.

Using class provides greater abstraction and follows the OOP convention.

Thiolk committed:
Revert "Implement Task class in Level-1 code"

This reverts commit 76bdf215b2b73838aff7a6d2e6c0d70fa72d490c.

Thiolk committed:
Implement Task class in Level-2 code

Methods were used to handle the operations of tasks.

Task class is created and used to manage the operations of task.

Using class provides greater abstraction and follows the OOP convention.

Thiolk committed:
Replace String with Task to fix bug

Thiolk committed:
Add extra checkers in checkInput method

Current bot can only check for list or bye.

To fulfil Level-3 requirements, more checks need to be added.

Additional if else statements were added.

Thiolk committed:
Update listTasks Method

Current listTasks cannot show whether task is completed.

Add an if else statement to check if task is completed
and print corresponding output.

Thiolk committed:
Shift tasks related methods from main class to task class

Methods related to Tasks should not be kept in main.

Move tasks related methods from Main.java to Task.java.

This better follows the convention of OOP.

Thiolk committed:
Replace if else statements with switch

Current code uses if else statements to help bot make a decision.

Replace if else statements with switch.

The if else chunk can grow to become very large.
Using switch can increase the readability of the code.

Thiolk committed:
Complete requirements for level 3

Bot is unable to print correct output for user input "mark" and "unmark".

Add correct outputs to the code and use setStatus method.

Thiolk committed:
Improve readability and refactoring

Current code has magic numbers and duplicate methods that
accomplish the same task of checking for input
It was also not at the same level of abstraction

Consolidate Duplicate Conditional Fragments by removing
the switch statement in the main method. This also removed the
magic numbers that were in the main method.
Refactor code by extracting method as the println statements
from the marking and unmarking of tasks can be grouped together.

This is done according to the SE best practices of improving
readability and refactoring

Thiolk committed:
Implement Todo Class which is a child of Task Class

To meet Level-4 requirements, a Todo class is needed to
handle Todo type tasks.

Create a Todo Class as a child of the Task Class as it is a
subset of the Task Class.

This follows OOP best practices of inheritance and
polymorphism.

Thiolk committed:
Implement Deadline Class which is a child of Task Class

To meet Level-4 requirements, a Deadline class is needed to
handle Deadline type tasks.

Create a Deadline Class as a child of the Task Class as it is a
subset of the Task Class.

This follows OOP best practices of inheritance and
polymorphism.

Thiolk committed:
Implement Event Class which is a child of Task Class

To meet Level-4 requirements, a Event class is needed to
handle Event type tasks.

Create a Event Class as a child of the Task Class as it is a
subset of the Task Class.

This follows OOP best practices of inheritance and
polymorphism.

Thiolk committed:
Refactor code using Extract Method

Current code has 3 println statements per task type.

Refactor code using Extract method by putting the 3 println statements
into a separate method whose name explains the purpose of
the println statements.

Improves the quality of the code by improving code structure.

Thiolk committed:
Allow Duke to handle exceptions

Duke is operating in an ideal world where there are no invalid inputs.

Users might not input commands in the correct format and Duke
needs to be able to handle it these exceptions

Add Duke Exceptions class to handle the various exceptions that can
occur when a user uses Duke. Duke will print an error message when
user inputs a task type that it does not recognise.

Thiolk committed:
Handle exceptions related to the todo task

Duke is unable to handle incorrect user commands related to the
different task types. if the command is not formatted correctly, Duke
will crash.

Add exceptions to check that user has inputted the name of the todo
task in the correct format.

Toh-HongFeng commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,62 @@
+import java.util.Scanner;
+import java.util.ArrayList;
 public class Duke {
+    public static void printHorizontalLine() {
```
Method names can be changed to PascalCase

Thiolk replied:
I believe that names representing methods must be verbs and written in camelCase.

Toh-HongFeng commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+public class Deadline extends Task{
+    protected String by;
+
+    public Deadline (String name, boolean isCompleted, String by) {
```
Can standardize across files to have no space between method name and open parenthesis.

Thiolk replied:
thank you

Toh-HongFeng commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+public class Deadline extends Task{
+    protected String by;
```
Good variable naming.

Toh-HongFeng commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,62 @@
+    public static void checkInput(String input, ArrayList<Task> listOfTasks) {
+        String[] inputs = input.split(" ");
+        switch (inputs[0]) {
+        case "bye":
```
Correct indentation for flow control.

Toh-HongFeng commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,62 @@
+        String input;
+        ArrayList<Task> listOfTasks = new ArrayList<>();
+        do {
+            Scanner in = new Scanner(System.in);
```
Variable name can be less vague.

Toh-HongFeng reviewed (commented):
Good coding standards.

jeromeongithub commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,120 @@
+        int nameEndIndex = input.indexOf("/by");
+        String name = input.substring(nameStartIndex + 1, nameEndIndex -1);
+        //check name length
+        String dueDate = input.substring(nameEndIndex + 4);
```
I noticed that you used "magic numbers" like 4 (in this line) and 6 (in line 82). Perhaps you could try processing the inputs in a different way? Personally, I found the String.split() function to be very useful for doing so.

jeromeongithub commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,62 @@
+
     public static void main(String[] args) {
         String logo = " ____        _        \n"
                 + "|  _ \\ _   _| | _____ \n"
                 + "| | | | | | | |/ / _ \\\n"
                 + "| |_| | |_| |   <  __/\n"
                 + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        printHelloMessage(logo);
+        String input;
+        ArrayList<Task> listOfTasks = new ArrayList<>();
+        do {
+            Scanner in = new Scanner(System.in);
+            input = in.nextLine();
+            printHorizontalLine();
+            checkInput(input, listOfTasks);
```
It looks like there are many levels of abstraction within your main function. Perhaps you could try to keep it at a single level of abstraction (SLAP)? I found the lecture example of Contacts1 to be very useful as reference to apply SLAP to my own work.

Thiolk replied:
thank you for your comment, I have referenced the lecture and found it to be very useful

jeromeongithub commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,7 @@
+public class DukeException extends Exception {
+    protected String errorMsg;
```
I think "errorMessage" would be better than "errorMsg" here as we are required to avoid 'texting-style' spelling.

Thiolk replied:
thank you

jeromeongithub commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,62 @@
+            Task.listTasks(listOfTasks);
+            break;
+        case "mark":
+            Task markTask = listOfTasks.get(Integer.parseInt(inputs[1])-1);
```
Perhaps you could name "markTask" differently as a noun? The current naming makes it sound like a verb (mark the task) which is usually used to name methods.

Thiolk replied:
thank you

jeromeongithub reviewed (commented):
Great job overall! It was very easy to find things in your code, especially the methods because they were very explicitly named.

Thiolk committed:
Add delete function to Duke Chatbot

Duke chatbot is unable to delete tasks added previously.

Add method to delete task in Task class and add a case "delete"

Thiolk committed:
Revert "Add delete function to Duke Chatbot"

This reverts commit 94d5563f05a647fa7b4de91d0cc7e90d46e37012.

Thiolk committed:
Add delete function to Duke Chatbot

Duke chatbot is unable to delete tasks added previously.

Add method to delete task in Task class and add a case "delete"

Thiolk committed:
Improve code quality and standards

Thiolk committed:
Add read file function to Duke Chatbot

Duke is unable to save and read previous user inputs.
With initialiseDuke method, Duke can read previous user inputs
or create a new file to store inputs if this is the user's first time

Duke uses Java.io.File package for this function

Thiolk committed:
Allow Duke to save tasks inputted by user to hard disk

Duke can only read from duke.txt if it exists in hard disk
and add into its ArrayList.

Use File and FileWriter packages to allow Duke to save
user input into hard disk

Thiolk committed:
Merge branch 'branch-Level-6'

* branch-Level-6:
  Add delete function to Duke Chatbot

Thiolk committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Allow Duke to save tasks inputted by user to hard disk
  Add read file function to Duke Chatbot

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/Todo.java

resolve conflicts

Thiolk committed:
Stage manifest file

Thiolk committed:
Create more classes to focus on OOP

Implement classes Parser, Storage, TaskList and UI

Improves code quality and standard according to OOP guidelines

Thiolk committed:
Implement find function in Duke chat bot

Add method to find keyword that user inputted and
check if the keyword is in any task and print task if
present.

Thiolk committed:
Merge pull request #1 from Thiolk/branch-Level-9

Level-9

Thiolk committed:
Adds in JavaDoc comments

Thiolk committed:
Add in a user guide

Provides a manual for the users know how to
operate Duke ChatBot

Thiolk committed:
Update README.md

Add user guide

sevenseasofbri commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+    private TaskList listOfTasks;
+    private UI ui;
+
+    /**
```
Good use of JavaDoc!

sevenseasofbri commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+        } catch (IOException e) {
+            System.out.println("File not found");
+        }
+        ui.printGoodbyeMessage();
+        ui.printHorizontalLine();
```
Good use of OOP, assigning console operations to the Ui class !

sevenseasofbri commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,64 @@
+        String homePath = System.getProperty("user.home") + "/Desktop";
+        String folderPath = homePath + "/data";
+        String filePath = folderPath + "/duke.txt";
+        Duke ChatBot = new Duke(folderPath, filePath);
```
Since the object name is still technically a variable name, use camelCase and not PascalCase for naming it. In this case, it should be `chatBot` and not `ChatBot`. [See here](https://se-education.org/guides/conventions/java/basic.html#naming)

sevenseasofbri commented on src/main/java/DukeExceptionTest.java:
```diff
@@ -0,0 +1,8 @@
+    public static void checkMissingTodoName (String input) throws MissingTaskNameException{
+        String[] inputs = input.split(" ");
+        if (inputs.length < 2) {
+            throw new MissingTaskNameException("☹ OOPS!!! Please input the name of the todo task.");
```
Perhaps avoid using magic literals and store the string literal in a constant (static final var). [See here](https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-magic-numbers)

sevenseasofbri commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,104 @@
+     * @return List of tasks previously saved, else return empty list.
+     * @throws IOException If file is not found.
+     */
+    public ArrayList<Task> initialiseDuke()  throws IOException {
```
Maybe rename this function to reflect what it really does? Like `loadListData` or something along those lines. I think `initialiseDuke` may not describe this function correctly. [See here](https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-misleading-names)

sevenseasofbri reviewed (commented):
Good job Leng Kiat! 💯 Excellent job following the coding standards and readability guidelines. I have commented wherever necessary, but if you think any of my comments apply to any section of the code I haven't commented on, please exercise the changes accordingly.

Thiolk committed:
Add in JavaDoc comments for DukeExceptionTest

Thiolk committed:
Merge pull request #2 from Thiolk/branch-A-JavaDoc

Add in JavaDoc comments for DukeExceptionTest

Thiolk committed:
Allow Duke to save the status of the task and to update saved file when
the task list is updated

Thiolk committed:
Allow Duke ChatBot to update saved data when user uses the mark, unmark
or delete command

Thiolk committed:
Change saved file path to fix bug where the file path cannot be found on
a windows computer

Thiolk committed:
Fix bugs in Duke ChatBot
